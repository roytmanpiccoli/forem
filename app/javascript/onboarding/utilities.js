export const jsonToForm = (data) => {
  const form = new FormData();
  data.forEach((item) => form.append(item.key, item.value));
  return form;
};

export const getContentOfToken = (token) =>
  document.querySelector(`meta[name='${token}']`).content;

export const updateOnboarding = (lastPage) => {
  const csrfToken = getContentOfToken('csrf-token');
  fetch('/onboarding_update', {
    method: 'PATCH',
    headers: {
      'X-CSRF-Token': csrfToken,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ user: { last_onboarding_page: lastPage } }),
    credentials: 'same-origin',
  });
};

/**
 * A util function to fetch the user's data from off of the document's body.
 *
 *
 * @returns {Object} A JSON object with the parsed user data.
 */
export const userData = () => {
  const { user = null } = document.body.dataset;
  return JSON.parse(user);
};
