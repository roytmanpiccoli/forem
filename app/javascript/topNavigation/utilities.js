function closeHeaderMenu(memberMenu, menuNavButton) {
  menuNavButton.setAttribute('aria-expanded', 'false');
  memberMenu.classList.remove('desktop', 'showing');
  delete memberMenu.dataset.clicked;
}

const firstItem = document.getElementById('first-nav-link');

function openHeaderMenu(memberMenu, menuNavButton) {
  menuNavButton.setAttribute('aria-expanded', 'true');
  memberMenu.classList.add('showing');

  if (!firstItem) {
    return;
  }

  // Focus on the first item in the menu
  (function focusFirstItem() {
    if (document.activeElement === firstItem) {
      // The first element has focus
      return;
    }

    firstItem.focus();
    // requestAnimationFrame is faster and more reliable than setTimeout
    // https://swizec.com/blog/how-to-wait-for-dom-elements-to-show-up-in-modern-browsers
    window.requestAnimationFrame(focusFirstItem);
  })();
}

/**
 * Determines whether or not a device is a touch device.
 *
 * @returns true if a touch device, otherwise false.
 */
export function isTouchDevice() {
  return /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini|DEV-Native-ios/i.test(
    navigator.userAgent,
  );
}

/**
 * Initializes the member navigation menu events.
 *
 * @param {HTMLElement} memberTopMenu The member menu in the top navigation.
 * @param {HTMLElement} menuNavButton The button to activate the member navigation menu.
 */
export function initializeMemberMenu(memberTopMenu, menuNavButton) {
  // Typically using CSS for hovering for the menu is the way to go. But... since we use InstantClick for
  // loading pages, the top header navigation never changes in terms of the DOM references.
  // Because of this, we're using mouse events to mouseover/mouseout on the member's avatar
  // to attach styles to get it to show the menu so that this works on desktop and mobile.
  if (navigator.userAgent === 'DEV-Native-ios') {
    document.body.classList.add('dev-ios-native-body');
  }
  const { classList } = memberTopMenu;
  menuNavButton.addEventListener('click', (_event) => {
    if (classList.contains('showing') && memberTopMenu.dataset.clicked) {
      closeHeaderMenu(memberTopMenu, menuNavButton);
      menuNavButton.focus();
    } else {
      openHeaderMenu(memberTopMenu, menuNavButton);
      memberTopMenu.dataset.clicked = 'clicked';
    }
  });

  if (isTouchDevice()) {
    memberTopMenu.addEventListener('focus', (_event) => {
      menuNavButton.setAttribute('aria-expanded', 'true');
    });
  } else {
    memberTopMenu.addEventListener('keyup', (e) => {
      if (e.key === 'Escape' && classList.contains('showing')) {
        closeHeaderMenu(memberTopMenu, menuNavButton);
        menuNavButton.focus();
      }
    });
  }

  memberTopMenu
    .querySelector('.crayons-header__menu__dropdown')
    .addEventListener('click', (event) => {
      // There is a click event listener on the body and we do not want
      // this click to be caught by it
      event.stopPropagation();

      // Close the menu if the user clicked or touched on mobile a link in the menu.
      closeHeaderMenu(memberTopMenu, menuNavButton);
      menuNavButton.focus();
    });

  document.addEventListener('click', (event) => {
    if (event.target.closest('#member-menu-button') === menuNavButton) {
      // The menu navigation button manages it's own click event.
      return;
    }

    // Close the menu if the user clicked or touched on mobile a link in the menu.
    closeHeaderMenu(memberTopMenu, menuNavButton);
  });

  const secondToLastNavLink = document.getElementById('second-last-nav-link');

  document
    .getElementById('last-nav-link')
    .addEventListener('blur', (_event) => {
      // When we tab out of the last link in the member menu, close
      // the menu.
      setTimeout(() => {
        if (document.activeElement === secondToLastNavLink) {
          return;
        }

        closeHeaderMenu(memberTopMenu, menuNavButton);
      }, 10);
    });
}

function toggleBurgerMenu() {
  const { leftNavState = 'closed' } = document.body.dataset;
  document.body.dataset.leftNavState =
    leftNavState === 'open' ? 'closed' : 'open';
}

function showMoreMenu({ target }) {
  target.nextElementSibling.classList.remove('hidden');
  target.classList.add('hidden');
}

/**
 * Gets a reference to InstantClick
 *
 * @param {number} [waitTime=2000] The amount of time to wait
 * until giving up waiting for InstantClick to exist
 *
 * @returns {Promise<object>} The global instance of InstantClick.
 */
export async function getInstantClick(waitTime = 2000) {
  return new Promise((resolve, reject) => {
    const failTimer = setTimeout(() => {
      clearInterval(timer);
      reject(new Error('Unable to resolve InstantClick'));
    }, waitTime);

    const timer = setInterval(() => {
      if (typeof InstantClick !== 'undefined') {
        clearTimeout(failTimer);
        clearInterval(timer);
        resolve(InstantClick);
      }
    });
  });
}

/**
 * Initializes the hamburger menu for mobile navigation
 *
 * @param {HTMLElement[]} menuTriggers
 * @param {HTMLElement[]} moreMenus
 */
export function initializeMobileMenu(menuTriggers, moreMenus) {
  menuTriggers.forEach((trigger) => {
    trigger.onclick = toggleBurgerMenu;
  });

  moreMenus.forEach((trigger) => {
    trigger.onclick = showMoreMenu;
  });
}

/**
 * Sets the icon link visually for the current page if the current page
 * is one of the main icon links of the top navigation.
 *
 * @param {string} currentPage
 * @param {[string, HTMLElement][]} pageEntries
 */
export function setCurrentPageIconLink(currentPage, pageEntries) {
  pageEntries
    // Filter out nulls (means the user is logged out so most icons are not in the logged out view)
    .filter(([, iconLink]) => iconLink)
    .forEach(([page, iconLink]) => {
      if (currentPage === page) {
        iconLink.blur();
        iconLink.setAttribute('aria-current', 'page');
      } else {
        iconLink.removeAttribute('aria-current');
      }
    });
}
