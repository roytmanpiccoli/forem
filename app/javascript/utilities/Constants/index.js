export const chatDrawPalettes = [
  '#F4908E',
  '#F2F097',
  '#88B0DC',
  '#F7B5D1',
  '#53C4AF',
  '#FDE38C',
];
