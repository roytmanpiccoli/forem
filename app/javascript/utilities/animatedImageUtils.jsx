import { h, render } from 'preact';
import Freezeframe from 'freezeframe';
import { Icon, ButtonNew as Button } from '@crayons';
import Play from '@images/play.svg';
import Pause from '@images/pause.svg';

const toggleGifButtonPressedState = (button) => {
  const currentlyPressed = button.getAttribute('aria-pressed') === 'true';
  button.setAttribute('aria-pressed', !currentlyPressed);
};

/**
 * Helper function that will initialize play/pause functionality for any image with the attribute data-animated="true"
 */
export const initializePausableAnimatedImages = (animatedImages = []) => {
  if (animatedImages.length > 0) {
    const freezeframes = [];

    for (let i = 0; i < animatedImages.length; i++) {
      const image = animatedImages[i];
      // Give the image an ID so that we can uniquely target it in freezeframe
      image.setAttribute('id', `animated-${i}`);

      // Remove the surrounding links for the image, so it can be clicked to play/pause
      image.closest('a').outerHTML = image.outerHTML;

      freezeframes.push(
        new Freezeframe({
          selector: `img[id='animated-${i}']`,
          responsive: false,
          trigger: 'click',
        }),
      );
    }

    const handleFreezeframeReadyState = (_mutationList, observer) => {
      // Check if freezeframe has finished initializing
      const initializedFrames = document.querySelectorAll(
        '.ff-container.ff-ready',
      );

      if (initializedFrames.length < freezeframes.length) {
        // Not ready yet, do nothing
        return;
      }

      // Freezeframes are ready, and we can disconnect our observer
      observer.disconnect();

      // Freezeframes are "paused" by default. If a user's settings allow, we immediately restart the animation.
      const okWithMotion = window.matchMedia(
        '(prefers-reduced-motion: no-preference)',
      ).matches;

      freezeframes.forEach((ff) => {
        if (okWithMotion) {
          ff.start();
        }

        // Freezeframe doesn't allow gifs to be stopped by keyboard press, so we add a button to handle it
        const ffWrapper = ff.items[0]['$container'];

        // We want to update the button's state when the gif is paused by image/canvas click
        ffWrapper.addEventListener('click', ({ currentTarget }) =>
          toggleGifButtonPressedState(
            currentTarget.querySelector('.gif-button'),
          ),
        );

        render(
          <Button
            aria-label="Pause animation playback"
            aria-pressed={!okWithMotion}
            className="gif-button fs-s gap-2"
            onClick={() => ff.toggle()}
          >
            <Icon src={Play} className="gif-play" />
            <Icon src={Pause} className="gif-pause" />
            GIF
          </Button>,
          ffWrapper,
        );
      });
    };

    // Freezeframe initializes asynchronously, but unfortunately doesn't offer a callback for when it's "ready".
    // This mutation observer allows us to complete some final tasks when it's complete.
    const readyWatcher = new MutationObserver(handleFreezeframeReadyState);
    readyWatcher.observe(document.querySelector('main'), {
      subtree: true,
      attributes: true,
      attributeFilter: ['class'],
    });
  }
};
