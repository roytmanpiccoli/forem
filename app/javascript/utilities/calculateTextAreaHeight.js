// These styles are applied to the hidden element we use to measure the height.
// !important styles are used to ensure no matter what style properties are attached to the given textarea, the hidden textarea will never become visible or cause layout jumps
const HIDDEN_TEXTAREA_STYLE = `
min-height:0 !important;
max-height:none !important;
height:0 !important;
visibility:hidden !important;
overflow:hidden !important;
position:absolute !important;
z-index:-1000 !important;
top:0 !important;
right:0 !important
`;

const SIZING_STYLE = [
  'letter-spacing',
  'line-height',
  'padding-top',
  'padding-bottom',
  'font-family',
  'font-weight',
  'font-size',
  'text-rendering',
  'text-transform',
  'width',
  'text-indent',
  'padding-left',
  'padding-right',
  'border-width',
  'box-sizing',
];

let hiddenTextarea;

/**
 * Helper function to get the height of the textarea based on the current text content
 *
 * @param {HTMLElement} uiTextNode The textarea to measure height of
 *
 * @returns {{height: number}} Object with the calculated height
 */
export const calculateTextAreaHeight = (uiTextNode) => {
  if (!hiddenTextarea) {
    hiddenTextarea = document.createElement('textarea');
    document.body.appendChild(hiddenTextarea);
  }

  // Copy all CSS properties that have an impact on the height of the content in
  // the textbox
  const {
    paddingSize,
    borderSize,
    boxSizing,
    sizingStyle,
  } = calculateNodeStyling(uiTextNode);

  // Need to have the overflow attribute to hide the scrollbar otherwise
  // text-lines will not calculated properly as the shadow will technically be
  // narrower for content
  hiddenTextarea.setAttribute(
    'style',
    `${sizingStyle};${HIDDEN_TEXTAREA_STYLE}`,
  );
  hiddenTextarea.value = uiTextNode.value || uiTextNode.placeholder || 'x';

  const baseHeight = hiddenTextarea.scrollHeight;

  if (boxSizing === 'border-box') {
    // border-box: add border, since height = content + padding + border
    return { height: baseHeight + borderSize };
  } else if (boxSizing === 'content-box') {
    // remove padding, since height = content
    return { height: baseHeight - paddingSize };
  }

  return { height: baseHeight };
};

const calculateNodeStyling = (node) => {
  const style = window.getComputedStyle(node);

  const boxSizing =
    style.getPropertyValue('box-sizing') ||
    style.getPropertyValue('-moz-box-sizing') ||
    style.getPropertyValue('-webkit-box-sizing');

  const paddingSize =
    parseFloat(style.getPropertyValue('padding-bottom')) +
    parseFloat(style.getPropertyValue('padding-top'));

  const borderSize =
    parseFloat(style.getPropertyValue('border-bottom-width')) +
    parseFloat(style.getPropertyValue('border-top-width'));

  const sizingStyle = SIZING_STYLE.map(
    (name) => `${name}:${style.getPropertyValue(name)}`,
  ).join(';');

  return {
    sizingStyle,
    paddingSize,
    borderSize,
    boxSizing,
  };
};
