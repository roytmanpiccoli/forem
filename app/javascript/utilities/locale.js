import I18n from "i18n-js"
const translationsDiv = document.getElementById('i18n-translations')
if (translationsDiv) {
  I18n.translations = JSON.parse(translationsDiv.dataset.translations);
}
I18n.defaultLocale = 'en';
I18n.locale = document.body.dataset.locale;
export function locale(term) {
  return I18n.t(term);
}