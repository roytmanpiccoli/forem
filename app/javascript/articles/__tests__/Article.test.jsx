/* eslint-disable no-irregular-whitespace */
import { h } from 'preact';
import { render } from '@testing-library/preact';
import { axe } from 'jest-axe';
import '@testing-library/jest-dom';
import { Article } from '..';
import { locale } from '../../utilities/locale';
import {
  article,
  articleWithOrganization,
  articleWithSnippetResult,
  articleWithReactions,
  videoArticle,
  articleWithComments,
  podcastArticle,
  podcastEpisodeArticle,
  userArticle,
  assetPath,
} from './utilities/articleUtilities';
import '../../../assets/javascripts/lib/xss';
import '../../../assets/javascripts/utilities/timeAgo';

const commonProps = {
  commentsIcon: assetPath('comments-bubble.png'),
  videoIcon: assetPath('video-camera.svg'),
};

describe('<Article /> component', () => {
  it('should have no a11y violations for a standard article', async () => {
    const { container } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={article}
        currentTag="javascript"
      />,
    );
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('should have no a11y violations for a featured article', async () => {
    const { container } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        isFeatured
        article={article}
        currentTag="javascript"
      />,
    );
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('should render a standard article', () => {
    const { container, queryByAltText } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={article}
        currentTag="javascript"
      />,
    );

    expect(container.firstChild).not.toHaveClass('crayons-story--featured', {
      exact: false,
    });
    expect(queryByAltText('Emil99 profile')).toBeDefined();
  });

  it('should render a featured article', () => {
    const { container, queryByAltText } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        isFeatured
        article={article}
        currentTag="javascript"
      />,
    );

    expect(container.firstChild).toHaveClass('crayons-story--featured', {
      exact: false,
    });
    expect(queryByAltText('Emil99 profile')).toBeDefined();
  });

  it('should render a rich feed', () => {
    const tree = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        isFeatured
        feedStyle="rich"
        article={article}
        currentTag="javascript"
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render a featured article for an organization', () => {
    const { container, queryByAltText } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        isFeatured
        article={articleWithOrganization}
        currentTag="javascript"
      />,
    );

    expect(container.firstChild).toHaveClass('crayons-story--featured', {
      exact: false,
    });
    expect(queryByAltText('Web info-mediaries logo')).toBeDefined();
    expect(queryByAltText('Emil99 profile')).toBeDefined();
  });

  it('should render a featured article for a video post', () => {
    const { queryByTitle } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        isFeatured
        article={videoArticle}
        currentTag="javascript"
      />,
    );

    expect(queryByTitle(/video duration/i)).toBeDefined();
  });

  it('should render with an organization', () => {
    const { queryByAltText } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={articleWithOrganization}
        currentTag="javascript"
      />,
    );

    expect(queryByAltText('Web info-mediaries logo')).toBeDefined();
    expect(queryByAltText('Emil99 profile')).toBeDefined();
  });

  it('should render with a flare tag', () => {
    const { queryByText } = render(
      <Article {...commonProps} isBookmarked={false} article={article} />,
    );

    expect(queryByText('#javascript', { selector: 'span' })).toBeDefined();
  });

  it('should render with a snippet result', () => {
    const { queryByText } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={articleWithSnippetResult}
      />,
    );

    expect(
      queryByText(
        '…copying Rest withdrawal Handcrafted multi-state Pre-emptive e-markets feed...overriding RSS Fantastic Plastic Gloves invoice productize systemic Monaco…',
      ),
    ).toBeDefined();
  });

  it('should render with reactions', () => {
    const { getByTitle } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={articleWithReactions}
      />,
    );

    const reactions = getByTitle('Number of reactions');

    expect(reactions.textContent).toEqual(`232 ${locale('core.reaction')}s`);
  });

  it('should render with comments', () => {
    const { getByTitle } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={articleWithComments}
      />,
    );

    const comments = getByTitle('Number of comments');

    expect(comments.textContent).toEqual(`213 ${locale('core.comment')}s`);
  });

  it('should render with an add comment button when there are no comments', () => {
    const { queryByTestId } = render(
      <Article {...commonProps} isBookmarked={false} article={article} />,
    );

    expect(queryByTestId('add-a-comment')).toBeDefined();
  });

  it('should render as saved on reading list', () => {
    const { queryByText } = render(
      <Article {...commonProps} isBookmarked article={articleWithComments} />,
    );

    expect(queryByText('Saved', { selector: 'button' })).toBeDefined();
  });

  it('should render as not saved on reading list', () => {
    const { queryByText } = render(
      <Article {...commonProps} isBookmarked={false} article={article} />,
    );

    expect(queryByText('Save', { selector: 'button' })).toBeDefined();
  });

  it('should render a video article', () => {
    const { queryByTitle } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={videoArticle}
        currentTag="javascript"
      />,
    );

    expect(queryByTitle(/video duration/i)).toBeDefined();
  });

  it('should render a podcast article', () => {
    const { queryByAltText, queryByText } = render(
      <Article
        {...commonProps}
        isBookmarked={false}
        article={podcastArticle}
      />,
    );

    expect(queryByAltText('Rubber local')).toBeDefined();
    expect(queryByText('podcast', { selector: 'span' })).toBeDefined();
  });

  it('should render a podcast episode', () => {
    const { queryByText } = render(
      <Article isBookmarked={false} article={podcastEpisodeArticle} />,
    );

    expect(queryByText('podcast', { selector: 'span' })).toBeDefined();
  });

  it('should render a user article', () => {
    const { queryByText } = render(<Article article={userArticle} />);

    expect(queryByText('person', { selector: 'span' })).toBeDefined();
  });
});
