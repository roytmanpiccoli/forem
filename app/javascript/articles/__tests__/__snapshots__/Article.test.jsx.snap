// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> component should render a rich feed 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="crayons-story cursor-pointer crayons-story--featured"
        data-content-user-id="23289"
        id="featured-story-marker"
      >
        <a
          aria-labelledby="article-link-62407"
          class="crayons-story__hidden-navigation-link"
          href="/some-post/path"
        >
          Unbranded Home Loan Account
        </a>
        <div
          role="presentation"
        >
          <a
            class="crayons-story__cover"
            href="/some-post/path"
            style="background-image: url(/images/10.png);"
            title="Unbranded Home Loan Account"
          >
            <span
              class="hidden"
            >
              Unbranded Home Loan Account
            </span>
          </a>
          <div
            class="crayons-story__body"
          >
            <div
              class="crayons-story__top"
            >
              <div
                class="crayons-story__meta"
              >
                <div
                  class="crayons-story__author-pic"
                >
                  <a
                    class="crayons-avatar crayons-avatar--l"
                    href="/Emil99"
                  >
                    <img
                      alt="Emil99 profile"
                      class="crayons-avatar__image"
                      loading="lazy"
                      src="/images/10.png"
                    />
                  </a>
                </div>
                <div>
                  <div>
                    <a
                      class="crayons-story__secondary fw-medium m:hidden"
                      href="/Emil99"
                    >
                      Stella Macejkovic
                    </a>
                    <div
                      class="profile-preview-card relative mb-4 s:mb-0 fw-medium hidden m:inline-block"
                    >
                      <button
                        aria-controls="story-author-preview-content-62407"
                        aria-label="Stella Macejkovic profile details"
                        class="profile-preview-card__trigger fs-s p-1 crayons-btn crayons-btn--ghost -ml-1 -my-2"
                        id="story-author-preview-trigger-62407"
                      >
                        Stella Macejkovic
                      </button>
                      <div
                        class="profile-preview-card__content crayons-dropdown p-4 pt-0 branded-7"
                        data-repositioning-dropdown="true"
                        data-testid="profile-preview-card"
                        id="story-author-preview-content-62407"
                        style=""
                      >
                        <div
                          class="gap-4 grid"
                        >
                          <div
                            class="-mt-4"
                          >
                            <a
                              class="flex"
                              href="/Emil99"
                            >
                              <span
                                class="crayons-avatar crayons-avatar--xl mr-2 shrink-0"
                              >
                                <img
                                  alt=""
                                  class="crayons-avatar__image"
                                  loading="lazy"
                                  src="/images/10.png"
                                />
                              </span>
                              <span
                                class="crayons-link crayons-subtitle-2 mt-5"
                              >
                                Stella Macejkovic
                              </span>
                            </a>
                          </div>
                          <div
                            class="print-hidden"
                          >
                            <button
                              class="c-btn c-btn--primary follow-action-button follow-user w-100"
                              data-info="{\\"id\\":23289,\\"className\\":\\"User\\",\\"name\\":\\"Stella Macejkovic\\",\\"style\\":\\"full\\"}"
                              type="button"
                            >
                              Follow
                            </button>
                          </div>
                          <div
                            class="author-preview-metadata-container"
                            data-author-id="23289"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <a
                    class="crayons-story__tertiary fs-xs"
                    href="/some-post/path"
                  >
                    <time
                      datetime="Tue, 18 Feb 2020 14:59:24 GMT"
                    >
                      February 18
                       (just now)
                    </time>
                  </a>
                </div>
              </div>
            </div>
            <div
              class="crayons-story__indention"
            >
              <h3
                class="crayons-story__title"
              >
                <a
                  href="/some-post/path"
                  id="article-link-62407"
                >
                  <span>
                    Unbranded Home Loan Account
                  </span>
                </a>
              </h3>
              <div
                class="crayons-story__tags"
              >
                <a
                  class="crayons-tag crayons-tag--filled"
                  href="/t/javascript"
                  style="--tag-bg: #0000001a; --tag-prefix: #000000; --tag-bg-hover: #0000001a; --tag-prefix-hover: #000000;"
                >
                  <span
                    class="crayons-tag__prefix"
                  >
                    #
                  </span>
                  javascript
                </a>
                <a
                  class="crayons-tag"
                  href="/t/ruby"
                >
                  <span
                    class="crayons-tag__prefix"
                  >
                    #
                  </span>
                  ruby
                </a>
                <a
                  class="crayons-tag"
                  href="/t/go"
                >
                  <span
                    class="crayons-tag__prefix"
                  >
                    #
                  </span>
                  go
                </a>
              </div>
              <div
                class="crayons-story__bottom"
              >
                <div
                  class="crayons-story__details"
                >
                  <a
                    aria-label="Comments for post Unbranded Home Loan Account (0)"
                    class="crayons-btn crayons-btn--ghost crayons-btn--s crayons-btn--icon-left"
                    data-testid="add-a-comment"
                    href="/some-post/path#comments"
                  >
                    <svg
                      class="crayons-icon"
                      height="24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M10.5 5h3a6 6 0 110 12v2.625c-3.75-1.5-9-3.75-9-8.625a6 6 0 016-6zM12 15.5h1.5a4.501 4.501 0 001.722-8.657A4.5 4.5 0 0013.5 6.5h-3A4.5 4.5 0 006 11c0 2.707 1.846 4.475 6 6.36V15.5z"
                      />
                    </svg>
                    <span
                      class="inline s:hidden"
                    >
                      0
                    </span>
                    <span
                      class="hidden s:inline"
                    >
                      [missing "en.core.add_comment" translation]
                    </span>
                  </a>
                </div>
                <div
                  class="crayons-story__save"
                >
                  <small
                    class="crayons-story__tertiary mr-2"
                  >
                    1 min read
                  </small>
                  <button
                    class="crayons-btn crayons-btn--s crayons-btn--secondary"
                    data-initial-feed="true"
                    data-reactable-id="62407"
                    id="article-save-button-62407"
                    type="button"
                  >
                    Save
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="crayons-story cursor-pointer crayons-story--featured"
      data-content-user-id="23289"
      id="featured-story-marker"
    >
      <a
        aria-labelledby="article-link-62407"
        class="crayons-story__hidden-navigation-link"
        href="/some-post/path"
      >
        Unbranded Home Loan Account
      </a>
      <div
        role="presentation"
      >
        <a
          class="crayons-story__cover"
          href="/some-post/path"
          style="background-image: url(/images/10.png);"
          title="Unbranded Home Loan Account"
        >
          <span
            class="hidden"
          >
            Unbranded Home Loan Account
          </span>
        </a>
        <div
          class="crayons-story__body"
        >
          <div
            class="crayons-story__top"
          >
            <div
              class="crayons-story__meta"
            >
              <div
                class="crayons-story__author-pic"
              >
                <a
                  class="crayons-avatar crayons-avatar--l"
                  href="/Emil99"
                >
                  <img
                    alt="Emil99 profile"
                    class="crayons-avatar__image"
                    loading="lazy"
                    src="/images/10.png"
                  />
                </a>
              </div>
              <div>
                <div>
                  <a
                    class="crayons-story__secondary fw-medium m:hidden"
                    href="/Emil99"
                  >
                    Stella Macejkovic
                  </a>
                  <div
                    class="profile-preview-card relative mb-4 s:mb-0 fw-medium hidden m:inline-block"
                  >
                    <button
                      aria-controls="story-author-preview-content-62407"
                      aria-label="Stella Macejkovic profile details"
                      class="profile-preview-card__trigger fs-s p-1 crayons-btn crayons-btn--ghost -ml-1 -my-2"
                      id="story-author-preview-trigger-62407"
                    >
                      Stella Macejkovic
                    </button>
                    <div
                      class="profile-preview-card__content crayons-dropdown p-4 pt-0 branded-7"
                      data-repositioning-dropdown="true"
                      data-testid="profile-preview-card"
                      id="story-author-preview-content-62407"
                      style=""
                    >
                      <div
                        class="gap-4 grid"
                      >
                        <div
                          class="-mt-4"
                        >
                          <a
                            class="flex"
                            href="/Emil99"
                          >
                            <span
                              class="crayons-avatar crayons-avatar--xl mr-2 shrink-0"
                            >
                              <img
                                alt=""
                                class="crayons-avatar__image"
                                loading="lazy"
                                src="/images/10.png"
                              />
                            </span>
                            <span
                              class="crayons-link crayons-subtitle-2 mt-5"
                            >
                              Stella Macejkovic
                            </span>
                          </a>
                        </div>
                        <div
                          class="print-hidden"
                        >
                          <button
                            class="c-btn c-btn--primary follow-action-button follow-user w-100"
                            data-info="{\\"id\\":23289,\\"className\\":\\"User\\",\\"name\\":\\"Stella Macejkovic\\",\\"style\\":\\"full\\"}"
                            type="button"
                          >
                            Follow
                          </button>
                        </div>
                        <div
                          class="author-preview-metadata-container"
                          data-author-id="23289"
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <a
                  class="crayons-story__tertiary fs-xs"
                  href="/some-post/path"
                >
                  <time
                    datetime="Tue, 18 Feb 2020 14:59:24 GMT"
                  >
                    February 18
                     (just now)
                  </time>
                </a>
              </div>
            </div>
          </div>
          <div
            class="crayons-story__indention"
          >
            <h3
              class="crayons-story__title"
            >
              <a
                href="/some-post/path"
                id="article-link-62407"
              >
                <span>
                  Unbranded Home Loan Account
                </span>
              </a>
            </h3>
            <div
              class="crayons-story__tags"
            >
              <a
                class="crayons-tag crayons-tag--filled"
                href="/t/javascript"
                style="--tag-bg: #0000001a; --tag-prefix: #000000; --tag-bg-hover: #0000001a; --tag-prefix-hover: #000000;"
              >
                <span
                  class="crayons-tag__prefix"
                >
                  #
                </span>
                javascript
              </a>
              <a
                class="crayons-tag"
                href="/t/ruby"
              >
                <span
                  class="crayons-tag__prefix"
                >
                  #
                </span>
                ruby
              </a>
              <a
                class="crayons-tag"
                href="/t/go"
              >
                <span
                  class="crayons-tag__prefix"
                >
                  #
                </span>
                go
              </a>
            </div>
            <div
              class="crayons-story__bottom"
            >
              <div
                class="crayons-story__details"
              >
                <a
                  aria-label="Comments for post Unbranded Home Loan Account (0)"
                  class="crayons-btn crayons-btn--ghost crayons-btn--s crayons-btn--icon-left"
                  data-testid="add-a-comment"
                  href="/some-post/path#comments"
                >
                  <svg
                    class="crayons-icon"
                    height="24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M10.5 5h3a6 6 0 110 12v2.625c-3.75-1.5-9-3.75-9-8.625a6 6 0 016-6zM12 15.5h1.5a4.501 4.501 0 001.722-8.657A4.5 4.5 0 0013.5 6.5h-3A4.5 4.5 0 006 11c0 2.707 1.846 4.475 6 6.36V15.5z"
                    />
                  </svg>
                  <span
                    class="inline s:hidden"
                  >
                    0
                  </span>
                  <span
                    class="hidden s:inline"
                  >
                    [missing "en.core.add_comment" translation]
                  </span>
                </a>
              </div>
              <div
                class="crayons-story__save"
              >
                <small
                  class="crayons-story__tertiary mr-2"
                >
                  1 min read
                </small>
                <button
                  class="crayons-btn crayons-btn--s crayons-btn--secondary"
                  data-initial-feed="true"
                  data-reactable-id="62407"
                  id="article-save-button-62407"
                  type="button"
                >
                  Save
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
